patches/tModLoader/Terraria/ModLoader/ModPylon.cs

```diff
@@ -0,0 +1,178 @@
+﻿using Microsoft.Xna.Framework;
+using Terraria.GameContent;
+using Terraria.Map;
+
+namespace Terraria.ModLoader
+{
+	/// <summary>
+	/// Extension to <seealso cref="ModTile"/> that streamlines the process of creating a
+	/// modded Pylon. Has all of ModTile's hooks for customization, but additional hooks for
+	/// Pylon functionality.
+	/// </summary>
+	/// <remarks>
+	/// One of the key features of this class is the <b>ValidTeleportCheck</b> process. At first glance it can look a bit
+	/// messy, however here is a rough break-down of the call process to help you if you're lost:
+	/// <br></br>
+	/// 1) Game queries if the specified player is near a Pylon (<seealso cref="TeleportPylonsSystem.IsPlayerNearAPylon"/>)
+	/// <br></br>
+	/// 2) Assuming Step 1 has passed, game queries if the DESTINATION PYLON (the pylon the player CLICKED on the map) has enough NPCs nearby (NPCCount step)
+	/// <br></br>
+	/// 3) Assuming Step 2 has passed, game queries if there is ANY DANGER at ALL across the entire map, ignoring the lunar pillar event (AnyDanger step)
+	/// <br></br>
+	/// 4) Assuming Step 3 has passed, game queries if the DESTINATION PYLON is in the Lihzahrd Temple before Plantera is defeated.
+	/// <br></br>
+	/// 5) Assuming Step 4 has passed, game queries if the DESTINATION PYLON meets its biome specifications for whatever type of pylon it is (BiomeRequirements step)
+	/// <br></br>
+	/// 6) Regardless of all the past checks, if the DESTINATION PYLON is a modded one, <seealso cref="ValidTeleportCheck_DestinationPostCheck"/> is called on it.
+	/// <br></br>
+	/// 7) The game queries all pylons on the map and checks if any of them are in interaction distance with the player, and if so, checks Step 2 and 5 on them (NPCCount &amp; BiomeRequirements step)
+	/// <br></br>
+	/// 8) Given that Step 7 finds a valid nearby pylon that satisfied the conditions, if that nearby pylon is a modded one, <seealso cref="ValidTeleportCheck_NearbyPostCheck"/> is called on it.
+	/// <br></br>
+	/// 9) Any <seealso cref="GlobalPylon"/> instances run <seealso cref="GlobalPylon.PostValidTeleportCheck"/>.
+	/// </remarks>
+	public abstract class ModPylon : ModTile
+	{
+
+		/// <summary>
+		/// What type of Pylon this ModPylon represents.
+		/// </summary>
+		public byte PylonType {
```

Comment:
similar to above comments, we either have to use `TeleportPylonType` everywhere or no-where. Fortunately, `Enum.ToString` does work with values outside the defined list, they just print as numbers. Though the first modded one might end up printing as 'Count' which could be confusing (unless we start modded above Count)